packages/userstorage/src/browser/user-storage-service-filesystem.spec.ts

```diff
@@ -0,0 +1,168 @@
+// interface SavedFileInfos { content: string, uri: string }
+let files: { [key: string]: string; } = {};
+
+before(async () => {
+    testContainer = new Container();
+
+    /* Preference bindings*/
+    testContainer.bind(PreferenceService).toSelf().inSingletonScope();
+    testContainer.bind(PreferenceServer).to(MockPreferenceServer).inSingletonScope();
+
+    /* FS mocks and bindings */
+    testContainer.bind(FileSystemWatcherServer).to(MockFilesystemWatcherServer).inSingletonScope();
+    testContainer.bind(FileSystemWatcher).toDynamicValue(ctx => {
+        const server = ctx.container.get<FileSystemWatcherServer>(FileSystemWatcherServer);
+        const prefs = ctx.container.get<FileSystemPreferences>(FileSystemPreferences);
+        const watcher = new FileSystemWatcher(server, prefs);
+
+        sinon.stub(watcher, 'onFilesChanged').get(() =>
+            mockOnFileChangedEmitter.event
+        );
+
+        return watcher;
+
+    }).inSingletonScope();
+    testContainer.bind(FileSystemPreferences).toDynamicValue(ctx => {
+        const preferences = ctx.container.get(PreferenceService);
+        return createFileSystemPreferences(preferences);
+    });
+
+    /* Mock logger binding*/
+    testContainer.bind(ILogger).to(MockLogger);
+    /* Stub getCurrentUserHome to return test home directory */
+    testContainer.bind(FileSystem).toDynamicValue(ctx => {
+        const fs = new MockFilesystem();
+
+        sinon.stub(fs, 'getCurrentUserHome').callsFake(() => Promise.resolve(
+            {
+                uri: 'file://' + homeDir,
+                lastModification: 0,
+                isDirectory: true
+            }));
+
+        sinon.stub(fs, 'resolveContent').callsFake((uri): Promise<{ stat: FileStat, content: string }> => {
+            const content = files[uri];
+            return Promise.resolve(
+                { stat: { uri: uri, lastModification: 0, isDirectory: false }, content: content }
+            );
+        });
+
+        sinon.stub(fs, 'setContent').callsFake((filestat, content: string) => {
+            files[filestat.uri] = content;
+            return Promise.resolve(content);
+        });
+
+        sinon.stub(fs, 'getFileStat').callsFake(uri =>
+            Promise.resolve({ uri, lastModification: 0, isDirectory: false })
+        );
+
+        return fs;
+    }).inSingletonScope();
+
+    testContainer.bind(UserStorageService).to(UserStorageServiceFilesystemImpl).inSingletonScope();
+    userStorageService = testContainer.get<UserStorageServiceFilesystemImpl>(UserStorageService);
+});
+
+describe('User Storage Service (Filesystem implementation)', () => {
+    beforeEach(() => {
+        files = {};
+    });
+
+    it('Should return a user storage uri from a filesystem uri', () => {
+        const testFile = 'keymaps.json';
+
+        const test = UserStorageServiceFilesystemImpl.toUserStorageUri(userStorageFolder, new URI('file://' + homeDir + '/' + THEIA_USER_STORAGE_FOLDER + '/' + testFile));
```

Comment:
It looks strange that the client code should use static functions o `UserStorageServiceFilesystemImpl`. Does not it mean that to use another impl all clients should be changed?
If not then these functions should be moved to `UserStorageUri` since they are independent of impl.
If yes then they should be provided in more abstract manner.

Reply:
@akosyakov I think they're only to be used by the UserStorage fs impl, they do not really make sense as API, so how exactly should they be changed. Private static methods of the impl? I'm not exactly sure about this...

Reply:
ok, at some point APIs should be called by real clients, how do they construct `UserStorageUri`, something like `UserStorageUri.create(path)`?

Reply:
They should not transform a file URI, for sure. We can have a general `UserStorageUri.create(path)` though. Maybe in another PR.